piccolo_api/change_password/endpoints.py

```diff
@@ -0,0 +1,181 @@
+from __future__ import annotations
+
+import os
+import typing as t
+from abc import ABCMeta, abstractproperty
+from json import JSONDecodeError
+
+from jinja2 import Environment, FileSystemLoader
+from piccolo.apps.user.tables import BaseUser
+from starlette.endpoints import HTTPEndpoint, Request
+from starlette.exceptions import HTTPException
+from starlette.responses import (
+    HTMLResponse,
+    PlainTextResponse,
+    RedirectResponse,
+)
+from starlette.status import HTTP_303_SEE_OTHER
+
+from piccolo_api.session_auth.tables import SessionsBase
+from piccolo_api.shared.auth.styles import Styles
+
+if t.TYPE_CHECKING:  # pragma: no cover
+    from jinja2 import Template
+    from starlette.responses import Response
+
+
+CHANGE_PASSWORD_TEMPLATE_PATH = os.path.join(
+    os.path.dirname(os.path.dirname(__file__)),
+    "templates",
+    "change_password.html",
+)
+
+
+class ChangePasswordEndpoint(HTTPEndpoint, metaclass=ABCMeta):
+    @abstractproperty
+    def _auth_table(self) -> t.Type[BaseUser]:
+        raise NotImplementedError
+
+    @abstractproperty
+    def _change_password_template(self) -> Template:
+        raise NotImplementedError
+
+    @abstractproperty
+    def _styles(self) -> Styles:
+        raise NotImplementedError
+
+    def render_template(
+        self, request: Request, template_context: t.Dict[str, t.Any] = {}
+    ) -> HTMLResponse:
+        # If CSRF middleware is present, we have to include a form field with
+        # the CSRF token. It only works if CSRFMiddleware has
+        # allow_form_param=True, otherwise it only looks for the token in the
+        # header.
+        csrftoken = request.scope.get("csrftoken")
+        csrf_cookie_name = request.scope.get("csrf_cookie_name")
+
+        return HTMLResponse(
+            self._change_password_template.render(
+                csrftoken=csrftoken,
+                csrf_cookie_name=csrf_cookie_name,
+                request=request,
+                styles=self._styles,
+                **template_context,
+            )
+        )
+
+    async def get(
+        self, request: Request
+    ) -> t.Union[HTMLResponse, PlainTextResponse]:
+        session_user = (
+            await SessionsBase.select(SessionsBase.user_id)
+            .where(SessionsBase.token == request.cookies.get("id"))
+            .first()

```

Comment:
We can use `request.user.user` here instead.

Reply:
I used this to avoid use `AuthenticationMiddleware` and `SessionsAuthBackend` (as I understand correctly we need this middlewares to use `request.user.user`) because I want to protect only `change_password` endpoint not whole app. I need to see if I can change that.

Reply:
I think it will have to be used with authentication middleware.

Reply:
Ok. I will try to implement authentication middleware.